Hermes/appscale/hermes/stats/producers/cluster_stats.py

```diff
@@ -0,0 +1,315 @@
+    """
+    self._utc_timestamp_cursors = {}
+    self._local_cache = local_cache
+    self._include_lists = include_lists
+    self._limit = limit
+    self._fetch_latest_only = fetch_latest_only
+    self._last_debug = 0
+
+  @gen.coroutine
+  def get_current_async(self):
+    """ Implements StatsSource.get_current_async() method. Makes concurrent
+    asynchronous http calls to cluster nodes and collects new proxies stats.
+
+    Returns:
+      A Future object which wraps a dict with node IP as key and list of
+      new ProxiesStatsSnapshot as value
+    """
+    lb_ips = appscale_info.get_load_balancer_ips()
+    # Do multiple requests asynchronously and wait for all results
+    per_node_proxies_stats_dict = yield {
+      node_ip: self._new_proxies_stats_from_node_async(node_ip)
+      for node_ip in lb_ips
+    }
+    if time.time() - self._last_debug > CLUSTER_STATS_DEBUG_INTERVAL:
+      self._last_debug = time.time()
+      logging.debug(per_node_proxies_stats_dict)
+    raise gen.Return(per_node_proxies_stats_dict)
+
+  @gen.coroutine
+  def _new_proxies_stats_from_node_async(self, node_ip):
+    last_utc_timestamp = self._utc_timestamp_cursors.get(node_ip)
+    if node_ip == appscale_info.get_private_ip():
+      new_snapshots = self._local_cache.get_stats_after(last_utc_timestamp)
+      if self._limit is not None:
+        if self._fetch_latest_only and len(new_snapshots) > self._limit:
+          new_snapshots = new_snapshots[len(new_snapshots)-self._limit:]
+        else:
+          new_snapshots = new_snapshots[:self._limit]
+    else:
+      new_snapshots = yield _fetch_remote_stats_cache_async(
+        node_ip=node_ip, method_path='stats/local/proxies/cache',
+        fromdict_convertor=proxy_stats.proxies_stats_snapshot_from_dict,
+        last_utc_timestamp=last_utc_timestamp,
+        include_lists=self._include_lists, limit=self._limit,
+        fetch_latest_only=self._fetch_latest_only
+      )
+    if new_snapshots:
+      self._utc_timestamp_cursors[node_ip] = new_snapshots[-1].utc_timestamp
+    raise gen.Return(new_snapshots)
+
+
+@gen.coroutine
+def _fetch_remote_stats_cache_async(node_ip, method_path, fromdict_convertor,
+                                    last_utc_timestamp, limit, include_lists,
+                                    fetch_latest_only):
+  # Security header
+  headers = {SECRET_HEADER: options.secret}
+  # Build query arguments
+  arguments = {}
+  if last_utc_timestamp is not None:
+    arguments['last_utc_timestamp'] = last_utc_timestamp
+  if limit is not None:
+    arguments['limit'] = limit
+  if include_lists is not None:
+    arguments['include_lists'] = include_lists
+  if fetch_latest_only:
+    arguments['fetch_latest_only'] = True
+
+  url = "http://{ip}:{port}/{path}".format(
+    ip=node_ip, port=options.port, path=method_path)
+  request = httpclient.HTTPRequest(
+    url=url, method='GET', body=json.dumps(arguments), headers=headers,
```

Comment:
Is there a reason these arguments need to be in the body instead of as standard url variables?